libcloud/common/upcloud.py

```diff
@@ -0,0 +1,255 @@
+        """
+        return json.dumps(self.body)
+
+
+class UpcloudNodeDestroyer(object):
+    """
+    Helper class for destroying node.
+    Node must be first stopped and then it can be
+    destroyed
+
+    :param  upcloud_node_operations: UpcloudNodeOperations instance
+    :type   upcloud_node_operations: :class:`.UpcloudNodeOperations`
+
+    :param  sleep_func: Callable function, which sleeps.
+        Takes int argument to sleep in seconds (optional)
+    :type   sleep_func: ``function``
+
+    """
+
+    WAIT_AMOUNT = 2
+    SLEEP_COUNT_TO_TIMEOUT = 20
+
+    def __init__(self, upcloud_node_operations, sleep_func=None):
+        self._operations = upcloud_node_operations
+        self._sleep_func = sleep_func or time.sleep
+        self._sleep_count = 0
+
+    def destroy_node(self, node_id):
+        """
+        Destroys the given node.
+
+        :param  node_id: Id of the Node.
+        :type   node_id: ``int``
+        """
+        self._stop_called = False
+        self._sleep_count = 0
+        return self._do_destroy_node(node_id)
+
+    def _do_destroy_node(self, node_id):
+        state = self._operations.node_state(node_id)
+        if state == 'stopped':
+            self._operations.destroy_node(node_id)
+            return True
+        elif state == 'error':
+            return False
+        elif state == 'started':
+            if not self._stop_called:
+                self._operations.stop_node(node_id)
+                self._stop_called = True
+            else:
+                # Waiting for started state to change and
+                # not calling stop again
+                self._sleep()
+            return self._do_destroy_node(node_id)
+        elif state == 'maintenance':
+            # Lets wait maintenace state to go away and retry destroy
+            self._sleep()
+            return self._do_destroy_node(node_id)
+        elif state is None:  # Server not found any more
+            return True
+
+    def _sleep(self):
+        if self._sleep_count > self.SLEEP_COUNT_TO_TIMEOUT:
+            raise UpcloudTimeoutException("Timeout, could not destroy node")
+        self._sleep_count += 1
+        self._sleep_func(self.WAIT_AMOUNT)
+
+
+class UpcloudNodeOperations(object):
+    """
+    Helper class to start and stop node.
+
+    :param  conneciton: Connection instance
+    :type   connection: :class:`.UpcloudConnection`
+    """
+
+    def __init__(self, connection):
+        self.connection = connection
+
+    def stop_node(self, node_id):
+        """
+        Stops the node
+
+        :param  node_id: Id of the Node
+        :type   node_id: ``int``
+        """
+        body = {
+            'stop_server': {
+                'stop_type': 'hard'
+            }
+        }
+        self.connection.request('1.2/server/{0}/stop'.format(node_id),
+                                method='POST',
+                                data=json.dumps(body))
+
+    def node_state(self, node_id):
```

Comment:
Since this is a non-standard method name, please add the `ex_` prefix. And since it does not modify state, I think `ex_get_node_state` is better, but that's up to you.

Reply:
Yeah the get_node_state is a better name for it, but is the ex_ needed here because this is a helper class and not part of the UpcloudDriver api.

Reply:
Oh, right. Not needed, no, sorry.